### PR TITLE
Improve ipfailover configurability

### PR DESCRIPTION
**Pull request by pecameron**
Openshift 3.5 feature. 

Add options to 'oadm ipfailover' to configure the check script and
notify scripts and also control the period the check script runs.

Keepalived periodically checks whether the application is running properly.
In the default case the test is a simple verification that something is listening 
on the watch port. This PR permits the user to supply an additional check script
that is run in the ipfailover container context to verify that the application is
operating properly. For example, a web server can be tested by accessing the 
watch port and verifying the response.

Whenever a node changes state to MASTER, BACKUP, or FAULT a notify
script can be called. This script has 3 parameters filled in by keepalived:
$1 - "GROUP"|"INSTANCE"
$2 - name of group or instance
$3 - target state of transition ("MASTER"|"BACKUP"|"FAULT")

--check-script="check_script"
The check script is a script in the keepalived container that verifies
the service is running properly. The script must return 0 for OK and 1 for FAIL.
These checks are in addition to verifying that the watch port is listening.

--notify-script="notify_script"
The notify script is a script in the keepalived container that is
called whenever the keepalived state transitions to (MASTER|BACKUP|FAULT)

--check-interval=<n>
The check script is run every <n> seconds. Default is 2.

Note: the scripts name is the full path to the script.

Fixes bug 1362163
https://trello.com/c/228zu7Br/267-5-improve-the-configurability-of-the-ipfailover-container-operations

Signed-off-by: Phil Cameron pcameron@redhat.com

### CODE REVIEW

**Comment by smarterclayton** [comment]
Is this a check script for the other container, or for the processes in the keepalived container itself?

**Comment by pecameron** [comment]
@knobunc Here is the first draft. It covers the wants in the trello card. PTAL

**Comment by smarterclayton** [comment]
The descriptions in the PR aren't enough for me to understand what the scripts are supposed to do.  More detail in the doc.

File: images/ipfailover/keepalived/conf/settings.sh [hunk 1]
```diff
@@ -38,11 +38,35 @@ HA_REPLICA_COUNT=${OPENSHIFT_HA_REPLICA_COUNT:-"1"}
...
-#  same cluster.
+#  same cluster. Range 1..255
 #     HA_VRRP_ID_OFFSET=30
 #
 HA_VRRP_ID_OFFSET=${OPENSHIFT_HA_VRRP_ID_OFFSET:-"0"}
 
+# When the DC supplies an (non null) iptables chain
+# (OPENSHIFT_HA_IPTABLES_CHAIN) make sure the rule to pass keepalived
+# multicast (224.0.0.18) traffic is in the table.
+HA_IPTABLES_CHAIN=${OPENSHIFT_HA_IPTABLES_CHAIN:-""}
```
Comment (stevekuznetsov) [inline images/ipfailover/keepalived/conf/settings.sh]: On this line and all the others:
```sh
HA_IPTABLES_CHAIN="${OPENSHIFT_HA_IPTABLES_CHAIN:-""}"
                  ^                                  ^
```

File: images/ipfailover/keepalived/lib/config-generators.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ source "$(dirname "${BASH_SOURCE[0]}")/utils.sh"
 
 # Constants.
 readonly CHECK_SCRIPT_NAME="chk_${HA_CONFIG_NAME//-/_}"
-readonly CHECK_INTERVAL_SECS=2
+readonly CHECK_INTERVAL_SECS=${HA_CHECK_INTERVAL}
```
Comment (stevekuznetsov) [inline images/ipfailover/keepalived/lib/config-generators.sh]:
```sh
readonly CHECK_INTERVAL_SECS="${HA_CHECK_INTERVAL}"
                             ^                    ^
```

File: images/ipfailover/keepalived/lib/config-generators.sh [hunk 3]
```diff
@@ -59,6 +59,11 @@ function generate_script_config() {
   if [ "$port" = "0" ]; then
     echo "   script \"true\""
   else
+    if [[ -n $HA_CHECK_SCRIPT ]]; then
```
Comment (stevekuznetsov) [inline images/ipfailover/keepalived/lib/config-generators.sh]:
```sh
if [[ -n "${HA_CHECK_SCRIPT}" ]]; then
         ^ ^               ^^
```

File: images/ipfailover/keepalived/lib/config-generators.sh [hunk 4]
```diff
@@ -59,6 +59,11 @@ function generate_script_config() {
...
+      echo "   if [[ -f $HA_CHECK_SCRIPT ]]; then"
```
Comment (stevekuznetsov) [inline images/ipfailover/keepalived/lib/config-generators.sh]: Are you generating a script here? This is pretty insane. This expansion of `$HA_CHECK_SCRIPT` will happen at `echo` time, not at runtime.

File: images/ipfailover/keepalived/tests/verify_failover_image.sh [hunk 5]
```diff
@@ -1,5 +1,6 @@
 #!/bin/bash -e
 
+# set -x
```
Comment (stevekuznetsov) [inline images/ipfailover/keepalived/tests/verify_failover_image.sh]: remove

**Comment by pecameron** [comment]
@smarterclayton  - I added more to the description. See the trelo card and bug
https://trello.com/c/228zu7Br/267-5-improve-the-configurability-of-the-ipfailover-container-operations

File: images/ipfailover/keepalived/conf/settings.sh (hunk 1 above)
Comment (pecameron) [inline images/ipfailover/keepalived/conf/settings.sh]: @stevekuznetsov I shifted OPENSHIFT_HA_IPTABLES_CHAIN to here to be more consistent. 
I didn't create this file and am just adding a few fields. I am not sure how far to go with this.

This is work in progress. Thanks for your comments and I will modify the code as suggested. I am not sure how far I should go beyond adding this feature. Three is discussion on converting this to a go program and using the go template to generate the file. However, that is future work (if we decide to do it.)

The scripting here generates the config file for keepalived.

File: images/ipfailover/keepalived/lib/config-generators.sh (hunk 2 above)
Comment (pecameron) [inline images/ipfailover/keepalived/lib/config-generators.sh]: fixed

File: images/ipfailover/keepalived/lib/config-generators.sh (hunk 4 above)
Comment (pecameron) [inline images/ipfailover/keepalived/lib/config-generators.sh]: @stevekuznetsov This is creating a config file for keepalived. One of the parameters to keepalived is the name of a  check script that keepalived should call. The expansion must happen when the config file is created.

File: images/ipfailover/keepalived/tests/verify_failover_image.sh (hunk 5 above)
Comment (pecameron) [inline images/ipfailover/keepalived/tests/verify_failover_image.sh]: Changed.

**Comment by pecameron** [comment]
@stevekuznetsov Is there more information you need here? Are there any further changes needed?

**Comment by pecameron** [comment]
@stevekuznetsov You have "Changes requested". What additional changes would you like to have made?

**Comment by stevekuznetsov** [comment]
@pecameron There is one more comment about an IP address https://github.com/openshift/origin/pull/11644#pullrequestreview-8018316

I would also like to see if we can open an issue to investigate an alternative approach to this than generating a script from within a script.

**Comment by knobunc** [comment]
@stevekuznetsov Yeah, @pecameron already talked to me about the possibility of changing over to a template file, or something less horrific.  We opened the card https://trello.com/c/9w6pIkIU

File: images/ipfailover/keepalived/conf/settings.sh [hunk 6]
```diff
@@ -65,7 +89,7 @@ ADMIN_EMAILS=(${OPENSHIFT_HA_ADMIN_EMAILS:-"root@localhost"})
 EMAIL_FROM="ipfailover@openshift.local"
 
 #  IP address of the SMTP server.
-SMTP_SERVER=${OPENSHIFT_HA_SMTP_SERVER:-"127.0.0.1"}
+SMTP_SERVER="${OPENSHIFT_HA_SMTP_SERVER:-"127.0.0.A1"}"
```
Comment (pecameron) [inline images/ipfailover/keepalived/conf/settings.sh]: @stevekuznetsov Oops, I missed that. Its now fixed.

**Comment by pecameron** [comment]
@stevekuznetsov @knobunc This code is generating the keepalived config file. keepalived reads this file to configure itself. It may look like a script but it is the config file. Even if we move to a template, the resulting file will still look the same as this file. It will look like a script and it will still be the same config file.

OK, I am missing something here. Do we want to change keepalived to accept its configuration in a different format?

**Comment by knobunc** [comment]
@pecameron Longer-term, we probably want to move it to go with a template.  But that will be done as a feature later.

**Comment by pecameron** [comment]
@stevekuznetsov I made the changes. PTAL

**Comment by stevekuznetsov** [comment]
I do admit I'm not well versed on this but diffs like [this](https://github.com/openshift/origin/pull/11644/files#diff-44b90de30920b572c835eea63c5b535eR63) are clearly writing Bash code into a file somewhere. Whether that's a Bash script to be run later or a config file that allows for Bash inline... we're generating executable Bash code dynamically. I feel like that is a very fragile approach, and, yes, it should be something that we think about later, it does not block this PR.

**Comment by pecameron** [comment]
@knobunc There is 1 approved on this, Is there something to do before you can approve it as well? Is this  ready to merge?

**Comment by knobunc** [comment]
@pecameron it is ready to merge.  I'm giving master until Monday, then I have a few to put in.

**Comment by knobunc** [comment]
[merge]

**Comment by knobunc** [comment]
[test]

**Comment by knobunc** [comment]
@pecameron:
```
=== BEGIN TEST CASE ===
test/cmd/router.sh:75: executing 'oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=7 --dry-run -o yaml' expecting success and text 'value: MY_CHAIN'
FAILURE after 1.528s: test/cmd/router.sh:75: executing 'oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=7 --dry-run -o yaml' expecting success and text 'value: MY_CHAIN': the output content test failed
```

**Comment by pecameron** [comment]
[test]

**Comment by knobunc** [comment]
Failed:
```
In suite "github.com/openshift/origin/test/cmd/router", test case "test/cmd/router.sh:75: executing 'oadm ipfailover --virtual-ips=\"1.2.3.4\" --check-interval=1177 --dry-run -o yaml' expecting success and text 'value" failed:
=== BEGIN TEST CASE ===
test/cmd/router.sh:75: executing 'oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' expecting success and text 'value: 1177'
FAILURE after 0.554s: test/cmd/router.sh:75: executing 'oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' expecting success and text 'value: 1177': the output content test failed
Standard output from the command:
```
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_check/8687/consoleText

File: test/cmd/router.sh [hunk 7]
```diff
@@ -72,6 +72,9 @@ os::cmd::expect_success_and_text 'oadm ipfailover --virtual-ips="1.2.3.4" --dry-
 os::cmd::expect_success_and_text 'oadm ipfailover --virtual-ips="1.2.3.4" --dry-run -o name' 'deploymentconfig/ipfailover'
 os::cmd::expect_success_and_text 'oadm ipfailover --virtual-ips="1.2.3.4" --dry-run -o yaml' '1.2.3.4'
 os::cmd::expect_success_and_text 'oadm ipfailover --virtual-ips="1.2.3.4" --iptables-chain="MY_CHAIN" --dry-run -o yaml' 'value: MY_CHAIN'
+os::cmd::expect_success_and_text 'oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' 'value: 1177'
```
Comment (stevekuznetsov) [inline test/cmd/router.sh]: Stronger checks here, if possible, would look like:

```sh
$ oadm ipfailver .... # create an API object
$ oc get <object> -o jsonpath='{.path.to.check-interval}' # expecting 1177
```

Not sure if this is possible or how the ipfailover configuration is serialized in etcd, but that sort of check is more declarative and less fragile than `grep`ping through some YAML.

**Comment by pecameron** [comment]
I don't get this. The yaml does have 1177 and it didn't match. Do you 
know what is wrong with this? Could it be the " " around the 1177?

FAILURE after 0.554s: test/cmd/router.sh:75: executing 'oadm ipfailover 
--virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' 
expecting success and text 'value: 1177': the output content test failed
Standard output from the command:
apiVersion: v1
items:
- apiVersion: v1
   kind: ServiceAccount
   metadata:
     creationTimestamp: null
     name: ipfailover
- apiVersion: v1
   kind: DeploymentConfig
   metadata:
     creationTimestamp: null
     labels:
       ipfailover: ipfailover
     name: ipfailover
   spec:
     replicas: 1
     strategy:
       resources: {}
       type: Recreate
     template:
       metadata:
         creationTimestamp: null
         labels:
           ipfailover: ipfailover
       spec:
         containers:
         - env:
           - name: OPENSHIFT_HA_CHECK_INTERVAL
             value: "1177" <<<<<<<<<<--------- It is here
           - name: OPENSHIFT_HA_CHECK_SCRIPT

**Comment by stevekuznetsov** [comment]
Yes, it must be exactly that -- you are looking for the literal text `value: 1177` in the YAML, but it does not exist. This is why I made the comment that I did about how to make this test more robust.

**Comment by stevekuznetsov** [comment]
In addition, you're not checking that it's the check interval value that is set to 1177, nor that it isn't set to 11775, or that not-value is set to 1177... etc

**Comment by pecameron** [comment]
Steve,
How do I code the line in the test?
oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=1177 --dry-run 
-o yaml' 'value: 1177'
would
oadm ipfailover --virtual-ips="1.2.3.4" --check-interval=1177 --dry-run 
-o yaml' 'value: "1177"'

What do you think?

phil

=== BEGIN TEST CASE ===
test/cmd/router.sh:75: executing 'oadm ipfailover 
--virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' 
expecting success and text 'value: 1177'
FAILURE after 0.554s: test/cmd/router.sh:75: executing 'oadm ipfailover 
--virtual-ips="1.2.3.4" --check-interval=1177 --dry-run -o yaml' 
expecting success and text 'value: 1177': the output content test failed
Standard output from the command:
apiVersion: v1
items:
...
- apiVersion: v1
   kind: DeploymentConfig
...
        - env:
           - name: OPENSHIFT_HA_CHECK_INTERVAL
             value: "1177" <<<<<---------
           - name: OPENSHIFT_HA_CHECK_SCRIPT
           - name: OPENSHIFT_HA_CONFIG_NAME

**Comment by stevekuznetsov** [comment]
Yes, the following would work:
```
os::cmd::expect_success_and_text "..." 'value: "1177"'
```

But, for all of the reasons above I still think it is much better to not do `--dry-run` but actually serialize the object to `etcd` and use a `jsonpath` to get the value you want, then check that and that only. See an example [in `routes.sh`](https://github.com/openshift/origin/blob/master/test/cmd/routes.sh#L28-L29).

**Comment by knobunc** [comment]
[test]

**Comment by knobunc** [comment]
[test] yum failed

**Comment by pecameron** [comment]
@knobunc Still need help with this...
Thanks...

**Comment by knobunc** [comment]
```
# cover github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/client/restclient
[WARNING] `go test` had the following output to stderr:
cover: internal error: block 268 overlaps block 270
	/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/client/restclient/request.go:#27803,#27927 /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/client/restclient/request.go:#27848,#27867
```

https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_check/8905/consoleText

**Comment by smarterclayton** [comment]
That's a spurious error.  Doesn't actually cause failure.

**Comment by pecameron** [comment]
So, what do we do now?

**Comment by knobunc** [comment]
[test]

**Comment by pecameron** [comment]
[test]

**Comment by knobunc** [comment]
`/go/src/github.com/openshift/origin/test/cmd/router.sh: line 82: unexpected EOF while looking for matching `''`

https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_check/9400/consoleText

**Comment by pecameron** [comment]
@knobunc are we ready to merge this?

**Comment by knobunc** [comment]
[merge]

**Comment by knobunc** [comment]
[merge]

**Comment by pecameron** [comment]
@knobunc This failed again, passed once before you did the merge and failed after. Maybe another merge will work...

**Comment by pecameron** [comment]
@knobunc https://bugzilla.redhat.com/show_bug.cgi?id=1362163 comment #6 is for the doc changes. The code is not yet in...

**Comment by knobunc** [comment]
[merge] last flaked with https://github.com/openshift/origin/issues/11016 (https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/12459/)

**Comment by knobunc** [comment]
[merge] last flaked with https://github.com/openshift/origin/issues/10228 (https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/9749/consoleText)

**Pull request merged**